File: 2021/150. Evaluate Reverse Polish Notation.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> numStack;
        for (string t : tokens) {
            if (t.length() == 1 && !isdigit(t[0])) {              
                int num2 = numStack.top();
                numStack.pop();
                int num1 = numStack.top();
                numStack.pop();
                int ans;
                switch(t[0]) {
                    case '+':
                        ans = num1 + num2;
                        break;
                    case '-':
                        ans = num1- num2;
                        break;
                    case '*':
                        ans = num1 * num2;
                        break;
                    case '/':
                        ans = num1 / num2;
                        break;
                    default:
                        break;
                }
                numStack.push(ans);
            } else {
                numStack.push(stoi(t));
            }
        }
        return numStack.top();
    }
};

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> numStack;
        for (string& t : tokens) {
            if (t == "+" || t == "-" || t == "*" || t == "/") {
                int num2 = numStack.top();
                numStack.pop();
                int num1 = numStack.top();
                numStack.pop();
                int tmp = 0;
                switch (t[0]) {
                    case '+':
                        tmp = num1 + num2;
                        break;
                    case '-':
                        tmp = num1 - num2;
                        break;
                    case '*':
                        tmp = num1 * num2;
                        break;
                    case '/':
                        tmp = num1 / num2;
                        break;                    
                }
                numStack.push(tmp);
            } else {
                numStack.push(stoi(t));
            }
        }
        return numStack.top();
    }
};

File: 2024/150. Evaluate Reverse Polish Notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        def cal(nums, op):
            num2 = nums[-1]
            nums.pop()
            num1 = nums[-1]
            nums.pop()
            if op == '+':
                nums.append(num1 + num2)
            elif op == '-':
                nums.append(num1 - num2)
            elif op == '*':
                nums.append(num1 * num2)
            elif op == '/':
                nums.append(int(num1 / num2))
        nums = list()
        for t in tokens:
            if t == '+' or t == '-' or t == '*' or t == '/':
                cal(nums, t)
            else:
                nums.append(int(t))
        return nums[-1]
